DS-2490: adds org.dspace.identifier.VersionedDOIIdentifierProvider

pnbecker opened the pull request:
https://jira.duraspace.org/browse/DS-2490

This PR adds a VersionedDOIIdentifierProvider for better handling of Item Level Versioning with DOIs. It also contains important bugfixes to make DOIs working again after the introduction of the Service Level API.

Depends on #900

pnbecker committed:
DS-2490: Add versioned DOIs and adapt DOI support to service API

pnbecker commented:
I have to test this again now, after service api was introduced.

pnbecker committed:
DataCiteConnector: don't rely on handles.

While the DOIIdentifierProvider checks our database to ensure that a DOI
is not reserverd/registered for another DSpaceObject.
The DataCiteConnector was checking this again agains  the meta data we
send to DataCite. Within these meta data we are submitting the handle as
alternative identifier and used it to check whether a DOI is
reserved/registered for a specific DSpaceObject. This commit removes
the checks from the DataCiteConnector for the following reasons:
 * In case we want to completly replace handles by DOIs one day, we
   shouldn't rely on handles. Further ongoing: an IdentifierProvider
   shouldn't rely on other identifiers than the one he provides.
 * If Item Level Versioning is enabled the handles may change depending
   on the VersionedHandleIdentifier that is used. This makes handles
   unreliable for the checks we were using them before.
   handles.

pnbecker committed:
DOIConumser: Ignore Workspace and Workflow items, reduce logging

pnbecker committed:
IdentifierServiceImpl.reserve(ctx, dso) should mint and reserve

pnbecker commented:
Rebased to resolve merge conflict.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/DOI.java:
```diff
@@ -67,11 +70,20 @@ public void setDoi(String doi) {
+    
+    public void setResourceTypeId(Integer resourceTypeId)
+    {
+        this.resourceTypeId = resourceTypeId;
```
I would leave out this setter & do it like ResourcePolicy does it: https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/authorize/ResourcePolicy.java#L170

Because having a separate setter for DSpaceObject & one for type will lead to issues.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:
```diff
@@ -768,68 +765,97 @@ public String getDOIByObject(Context context, DSpaceObject dso)
-                            DOIIdentifierException.DOI_ALREADY_EXISTS);
+                    // doi was deleted, check resource type
+                    if (doi.getResourceTypeId() != null
+                            && doi.getResourceTypeId() != dso.getType())
```
Not 100% sure if this "!=" will work in all cases, the getResourceTypeId() returns an object while dso.getType() is a primitive.
Can the resourceTypeId ever be null ? if not perhaps convert it to an int in the DOI class.

pnbecker replied:
This is well handled by Java's autoboxing feature as long as doi.getResourceTypeId() is not null, which is checked in the same if clause...

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+    @Autowired(required = true)
+    private VersioningService versioningService;
+    @Autowired(required = true)
+    private VersionHistoryService versionHistoryService;
```
Make this variable protected, so that if custom code is needed the class inheriting from this one can make use it.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+    private static final String pattern = "\\d+\\" + String.valueOf(DOT) +"\\d+";
+    
+    @Autowired(required = true)
+    private VersioningService versioningService;
```
Make this variable protected, so that if custom code is needed the class inheriting from this one can make use it.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+ /** log4j category */
+    private static Logger log = Logger.getLogger(VersionedDOIIdentifierProvider.class);
+    
+    private DOIConnector connector;
```
Make this variable protected, so that if custom code is needed the class inheriting from this one can make use it.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+    private DOIConnector connector;
+
+    static final char DOT = '.';
+    private static final String pattern = "\\d+\\" + String.valueOf(DOT) +"\\d+";
```
Make this variable protected, so that if custom code is needed the class inheriting from this one can make use it.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+        }
+        catch (SQLException ex)
+        {
+            log.error("Error while attemping to retrieve information about a DOI for " 
```
The actual exception isn't logged here.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+                try {
+                    loadOrCreateDOI(context, dso, versionedDOI);
+                } catch (SQLException ex) {
+                    throw new RuntimeException("A problem with the database connection occured.", ex);
```
Best to log SQLExceptions.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+        {
+            doi = loadOrCreateDOI(context, dso, doiIdentifier);
+        } catch (SQLException ex) {
+            log.error("Error in databse connection: " + ex.getMessage());
```
Best to log the exception.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java:
```diff
@@ -0,0 +1,350 @@
+        return identifier;
+    }
+    
+    private String getDOIPostfix(String identifier) 
```
Make this method protected, so that if custom code is needed the class inheriting from this one can make use it.

KevinVdV commented on dspace-api/src/main/java/org/dspace/identifier/doi/DOIConsumer.java:
```diff
@@ -18,6 +19,8 @@
 import org.dspace.identifier.IdentifierNotFoundException;
 import org.dspace.utils.DSpace;
+import org.dspace.workflow.factory.WorkflowServiceFactory;
+import org.dspace.xmlworkflow.state.Workflow;
```
Why is this dependency here, it appears unused (checking in github so not 100% sure)

pnbecker committed:
Improves inheritability and logging, removes DOI.setResourceType()

pnbecker commented:
Hi @KevinVdV,

I added a commit following all your suggestions except one. With Java's autoboxing it is no problem to compare an Integer and an int using == as long as the Integer is not null. I'm not 100% sure, but IIRC the field resource type is set null during the creation of a DOI. So I didn't wanted to change it to int.

I added every other suggestion you made. Thank you for the review!
  Pascal

pnbecker commented:
I forgot to mention, that I rebased the PR against the current master...

tdonohue commented:
@pnbecker : this looks to need a rebase

pnbecker commented:
@tdonohue thanks for pinging me, done.

KevinVdV commented:
I took another look at the code and all my issues have been resolved (haven't tested it).
